addons/crm/models/crm_lead.py

```diff
@@ -382,31 +388,56 @@ def create(self, vals_list):
         return leads
 
     def write(self, vals):
-        # stage change:
-        if 'stage_id' in vals:
-            stage_id = self.env['crm.stage'].browse(vals['stage_id'])
-            if stage_id.is_won:
-                vals.update({'probability': 100})
-
-        # stage change with new stage: update probability and date_closed
+        #   stage change with new stage: update probability and date_closed
         if vals.get('probability', 0) >= 100 or not vals.get('active', True):
             vals['date_closed'] = fields.Datetime.now()
         elif 'probability' in vals:
             vals['date_closed'] = False
 
+        if any(field in ['active', 'stage_id'] for field in vals):
+            vals = self._handle_won_lost(vals)
+
         write_result = super(Lead, self).write(vals)
         # Compute new automated_probability (and, eventually, probability) for each lead separately
         if self._should_update_probability(vals):
             self._update_probability()
 
         return write_result
 
+    def _handle_won_lost(self, vals):
+        # Global Won / Lost impacts
+        new_stage = self.env['crm.stage'].browse(vals.get('stage_id'))
+        if new_stage.is_won:
+            vals['probability'] = 100
+
+        # Lead by Lead impact
+        Lead = self.env['crm.lead']
+        leads_reach_won = Lead
```

Comment:
No "leads_leave_won" ?
Would be nice to explain that + general method purpose in docstring.